### DISCUSSION TITLE
Creating a page from migrations

### DISCUSSION THREAD

**Discussion by Xowap** [wagtail/wagtail#8340] [General]
I am currently working on converting some "Django admin" content (aka plain old Django models) into Wagtail pages. As such, I'd like to be able to create those Wagtail pages from a migration, because it would be pointless to do it manually and it's much simpler if I want to delete the old models right after they're migrated.

However, I guess that everyone here is aware of the complex tree structure in Wagtail as well as the unavailability of most of the code from migrations.

In order to go around this, I've sketched up the following utility class whose goal is to help creating content while being as future-proof as possible. More importantly, if it was included in a package (or in Wagtail directly) its abstraction would allow to add fixes when trouble shows up in newer version of treebard/wagtail.

Here is the monster, currently sitting inside a migration file:

```python
from typing import Any, Dict, Type

from django.apps import AppConfig
from django.conf import settings
from django.contrib.contenttypes.models import ContentType as RealContentType
from django.db import migrations, models
from treebeard.mp_tree import MP_Node
from wagtail.core import models as core_models
from wagtail.core.utils import get_supported_content_language_variant


class WagtailTreeMigrator:
    def __init__(self, apps: AppConfig):
        self.apps = apps

    def model(self, name: str) -> Type[models.Model]:
        return self.apps.get_model(name)

    def content_type(self, model: Type[models.Model]):
        ct = self.model("contenttypes.ContentType").objects.get_for_model(model)
        inject_base(ct, RealContentType)
        return ct

    def locale(self, language_code: str) -> core_models.Locale:
        return self.model("wagtailcore.Locale").objects.get(
            language_code=get_supported_content_language_variant(language_code)
        )

    def default_locale(self) -> core_models.Locale:
        return self.locale(settings.LANGUAGE_CODE)

    def site(self, is_default_site: bool = True, **kwargs) -> core_models.Site:
        return self.model("wagtailcore.Site").objects.get(
            is_default_site=is_default_site, **kwargs
        )

    def create_page(
        self, parent: core_models.Page, model: str, kwargs: Dict[str, Any]
    ) -> core_models.Page:
        cls = self.model(model)
        inject_base(parent, MP_Node)

        defaults = dict(
            content_type=self.content_type(cls),
            locale=self.default_locale(),
        )
        page = cls(**{**defaults, **kwargs})

        parent.add_child(instance=page)

        return page  # noqa


def inject_base(obj, cls):
    if cls not in obj.__class__.__bases__:
        obj.__class__.__bases__ = (cls,) + obj.__class__.__bases__


def forward(apps: AppConfig, _):
    m = WagtailTreeMigrator(apps)
    site = m.site()

    page = m.create_page(
        site.root_page,
        "home.YoloPage",
        dict(
            title="Yolo",
            slug="yolo",
        ),
    )
```

What are your thoughts on this?

- It's heretic and should probably be burned on a stick but does it produce a correct result? It seems to me that the pages are created properly.
- Would the community need this? I'm aware that my use case is a bit rare but I've come across several blog posts without definitive answers
- Is Wagtail interested in integrating something like this, in which case should I start polishing it in some way?

Thanks!
Rémy

**Comment by cnk** [comment]
A couple of thoughts:

1. I would not do this as a migration because I view it more like a one shot import rather than a reversible data transformation and I don't want this to be a step for all my dev/test database setups for all eternity. I would probably make a straight python class for this and then call it from a manage command. 
2. I would love to see this strategy written up and the code available on the blog post. But I feel like it's hard to make this general enough to include in Wagtail because everyone's data sources and page models are unique to their projects. If you made this sufficiently general purpose, you'll end up with something like [Drupal Migrate](https://www.drupal.org/project/migrate) - a project I find equal parts technically impressive and horrifying (because it's the only way to move between major versions of Drupal).

**Comment by nickmoreton** [comment]
I would agree with you @cnk 

This is something similar I was involved in recently...

I'd approach it by exposing the models in your Django app as json api and write a command/s to run and consume the api and process the results into Page models.

On a recent project importing from a Wordpress json API -> Django app -> Wagtail App

An issue we had was the Wordpress site is very large and the page tree very deep, the only data we had from Wordpress to recreate the page tree was a field with a parent page id on every page. We ended up importing every page at the top level (under home page) and keeping the Wordpress page id as part of the page model in Wagtail (script one). Then another script  moved the pages around after the import. (script two)

We had problems with duplicate slugs initially so created unique slugs in script one and saved the original slug in the Wagtail. page model so that in script two we could replace the slug with the original.

#### As a note 
the Django App isn't really required, we used it to do some other processing before importing into wagtail and it helped us understand the data coming out of Wordpress. But that was the API we used to import into Wagtail and not the Wordpress one.

**Comment by lb-** [comment]
Converting to a discussion as this is something that needs a lot more discovery before we end up putting something like this into Wagtail, if ever.
